crmsh/ui_resource.py

```diff
@@ -497,11 +498,67 @@ def do_constraints(self, context, resource):
         failcount <rsc> delete <node>
         failcount <rsc> show <node>"""
+        def sec_to_ms(s):
+            return s + '000'
+
+        def ms_to_sec(m):
+            return m[:len(m)-3]
+
+        if cmd == "set":
+            nodeid = utils.get_nodeid_from_name(node)
+            if nodeid is None:
+                context.fatal_error("Node {} not in this cluster".format(node))
+
+            # query the current failcount status
+            query_cmd = "cibadmin -Ql --xpath '/cib/status/node_state[@id='{}']'".format(nodeid)
+            rc, out, err = utils.get_stdout_stderr(query_cmd)
+            if rc != 0:
+                context.fatal_error(err)
+
+            # try to get failcount dict {operation:interval}
+            import re
+            failcount_res = re.findall(r'fail-count-{}#(.*)_([0-9]+)'.format(rsc), out)
+            if not failcount_res:
+                context.fatal_error("No failcount on node {} for resource {}".format(node, rsc))
+            failcount_dict = dict(failcount_res)
+
+            # validate for operation and interval
+            if operation and operation not in failcount_dict.keys():
+                context.fatal_error("Usage: failcount <rsc> set <node> <value> [operation] [interval]\n\
+                           Should specify operation between \"{}\"".format(' '.join(failcount_dict.keys())))
+            if (operation and interval) and (operation, sec_to_ms(interval)) not in failcount_res:
+                context.fatal_error("Usage: failcount <rsc> set <node> <value> [operation] [interval]\n\
+                           Should specify (operation, interval) between {}".\
+                           format([(op, ms_to_sec(inter)) for op, inter in failcount_res]))
+
+            # just one failcount entry
+            if len(failcount_res) == 1:
+                operation = failcount_res[0][0]
+                interval = failcount_dict[operation]
+                rsc = '{}#{}_{}'.format(rsc, operation, interval)
+
+            # multiple failcount entries for this resource and node
+            if len(failcount_res) > 1:
+                if operation and interval:
+                    rsc = '{}#{}_{}'.format(rsc, operation, sec_to_ms(interval))
+                elif int(value) == 0:
+                    # using '-P' option of 'crm_attribute' command
+                    cmd = "set_p"
+                    if operation:
+                        op_interval_str = '|'.join(['{}_{}'.format(operation, inter) for op, inter in failcount_res if op==operation])
+                    else:
+                        op_interval_str = '|'.join(['{}_{}'.format(op, inter) for op, inter in failcount_res])
+                    rsc = '{}#({})'.format(rsc, op_interval_str)
+                else:
+                    # value != 0
+                    context.fatal_error("Should specify (operation, interval) between {}".
+                            format([(op, ms_to_sec(inter)) for op, inter in failcount_res]))
```

Comment:
Thanks for the change. Of course like I mentioned, it would be more friendly with a logic like: "In the case of `len(failcount_res) > 1`, for a `non-zero value`, if `only one` fail-count entry is _matched_ with the specified `operation`, even if `interval` is not specified, just set it ".

Reply:
Great suggestion! I've changed like this:
```
                    # value != 0
                    if operation and len([op for op, _ in failcount_res if op == operation]) == 1:
                        rsc = '{}#{}_{}'.format(rsc, operation, failcount_dict[operation])
                    else:
                        context.fatal_error("Should specify (operation, interval) between {}".
                                format([(op, ms_to_sec(inter)) for op, inter in failcount_res]))
```